src/NetworkManagerUtils.c

```diff
@@ -67,85 +67,85 @@ nm_utils_get_shared_wifi_permission (NMConnection *connection)
 /*****************************************************************************/
 
 static char *
-get_new_connection_name (const GSList *existing,
+get_new_connection_name (const GSList *existing_connections,
                          const char *preferred,
                          const char *fallback_prefix)
 {
 	GSList *names = NULL;
 	const GSList *iter;
 	char *cname = NULL;
-	int i = 0;
-	gboolean preferred_found = FALSE;
+	guint i;
 
 	g_assert (fallback_prefix);
 
-	for (iter = existing; iter; iter = g_slist_next (iter)) {
+	for (iter = existing_connections; iter; iter = g_slist_next (iter)) {
 		NMConnection *candidate = NM_CONNECTION (iter->data);
 		const char *id;
 
 		id = nm_connection_get_id (candidate);
-		g_assert (id);
+		nm_assert (id);
+
 		names = g_slist_append (names, (gpointer) id);
 
-		if (preferred && !preferred_found && (strcmp (preferred, id) == 0))
-			preferred_found = TRUE;
+		if (   preferred
+		    && nm_streq (preferred, id)) {
+			/* the preferred name is already taken. Forget about it. */
+			preferred = NULL;
+		}
 	}
 
 	/* Return the preferred name if it was unique */
-	if (preferred && !preferred_found) {
+	if (preferred) {
 		g_slist_free (names);
 		return g_strdup (preferred);
 	}
 
 	/* Otherwise find the next available unique connection name using the given
 	 * connection name template.
 	 */
-	while (!cname && (i++ < 10000)) {
+	for (i = 1; TRUE; i++) {
 		char *temp;
-		gboolean found = FALSE;
 
 		/* Translators: the first %s is a prefix for the connection id, such
 		 * as "Wired Connection" or "VPN Connection". The %d is a number
 		 * that is combined with the first argument to create a unique
 		 * connection id. */
-		temp = g_strdup_printf (C_("connection id fallback", "%s %d"),
+		temp = g_strdup_printf (C_("connection id fallback", "%s %u"),
 		                        fallback_prefix, i);
 		for (iter = names; iter; iter = g_slist_next (iter)) {
-			if (!strcmp (iter->data, temp)) {
-				found = TRUE;
-				break;
+			if (nm_streq (iter->data, temp)) {
+				cname = temp;
+				goto found;
 			}
 		}
-		if (!found)
-			cname = temp;
-		else
-			g_free (temp);
+		g_free (temp);
```

Comment:
The search logic looks wrong here, after we find a match we should skip the generated name and try another.